Convert crm_simulate to formatted output

clumens opened the pull request:
This is a very large change by lines and number of files, but keep in mind a lot of that is caused by changing the expected cts-scheduler output.

kgaillot commented on tools/crm_simulate.c:
```diff
@@ -702,7 +700,8 @@ profile_one(const char *xml_file, long long repeat, pe_working_set_t *data_set,
     }
-    printf(" %.2f secs\n", (clock() - start) / (float) CLOCKS_PER_SEC);
+
+    out->message(out, "profile", xml_file, start, clock());
```
We now have an easy way of checking what the overhead of formatted output is, using pcmk_simtimes to compare before and after this PR

clumens replied:
```
$ tools/pcmk_simtimes /tmp/crm_simulate.before /tmp/crm_simulate.after
...
Averages: +0.01s  +2.03%
```

The bigger concerns I have are:

* Tons of this in /tmp/crm_simulate.after:  `Only 'private' parameters to stonith-fence_compute-fence-nova for unfencing compute-1 changed`
* Output to the screen seems to be batched up - it prints some, and then seems to hang for a long time before the next big chunk gets printed.

clumens replied:
Fixed the buffering.  Still have the harder problem to figure out.

clumens replied:
All that output is coming from https://github.com/ClusterLabs/pacemaker/pull/2322/commits/7d4c60091d69059597a83dcbe9e353423d467c72#diff-7b3f0ed3ea1b5a120ccc8da18b969eca9e1513b20db3d4665bb23e0efd02f2a7L283.  This didn't show up in crm_simulate profile output in the past because pe_flag_stdout was set after the profile function had run (and then the program exited).

So I'm not really sure how to fix this up so it only happens from crm_simulate sometimes.  We can't just use out->info or it'll always output (and get logged, ugh).  I really don't want to test fmt_name more than I already am, and I'm not sure if that would help in the first place.  Maybe pe_flag_stdout is more useful than we want to admit?

clumens replied:
I've updated 5e9de5b in a way that fixes this output problem, and I think it's a reasonable approach.

kgaillot commented on daemons/fenced/pacemaker-fenced.c:
```diff
@@ -1546,12 +1556,30 @@ main(int argc, char **argv)
     pcmk__serve_fenced_ipc(&ipcs, &ipc_callbacks);
 
+    pcmk__register_formats(NULL, formats);
+    rc = pcmk__output_new(&out, "log", NULL, argv);
```
Is the intent to use this in log_resource_details() instead of automatically creating one? Otherwise I don't see a point. Though looking at later commits, I wonder if you meant to drop this one

clumens replied:
The idea was that all the output messages in the scheduler would log instead of print to the screen when run via the daemons.

kgaillot replied:
That's a decent approach, and it means you don't need to create a separate logger in log_resource_details(), but it also means you can't use the existence of ->priv to distinguish crm_simulate from the daemons, and it means you need some sort of log level mechanism.

clumens replied:
Killing the separate logger shows new stuff getting added to the .scores files, so yeah definitely the right thing to do and also I clearly have some more work to do here.

kgaillot commented on lib/pacemaker/pcmk_sched_allocate.c:
```diff
@@ -1392,7 +1392,9 @@ stage5(pe_working_set_t * data_set)
-        dump_node_capacity(log_prio, "Original", node);
+        if (pcmk_is_set(data_set->flags, pe_flag_show_utilization)) {
+            out->message(out, "node-capacity", node, "Original");
+        }
```
The lost piece is being able to trace-log the message via the daemons

clumens replied:
I had hoped that using a log formatter here would allow the messages to be logged, though I see it would not be at the appropriate level.  I could potentially introduce a new message for this kind of thing, or even a whole new daemon-specific logger.  Trying to sort out all the different printing/logging needs of the daemons and command line tools is difficult.

I've also not tested the before and after to see what is actually happening.  I guess that's the next thing to do.

kgaillot replied:
The daemons don't set pe_flag_show_utilization, though they could if this custom message knew to use trace level when using the logger formatter.

One possibility would be to keep the log level in the log formatter's private data, and have a formatter-specific function to set it.

clumens replied:
Done - I've added a patch to allow setting log level and set the flags in the daemons so these output messages will be called.

kgaillot commented on include/pcmki/pcmki_output.h:
```diff
@@ -28,6 +28,8 @@ void pcmk__out_epilogue(pcmk__output_t *out, xmlNodePtr *xml, int retval);
 void pcmk__register_lib_messages(pcmk__output_t *out);
 
+void pcmk__register_sched_messages(pcmk__output_t *out);
+
```
Separating the messages into their own file is fine, but I'd rather have a single register function per library

clumens replied:
The problem I'm trying to solve here is that I originally intended lib/pacemaker/pcmk_output.c to be where output-related stuff to support development of libpacemaker lives.  That's what pcmk__out_prologue and pcmk__out_epilogue are doing there.  At some point, I put formatted output messages in there too because <lib>_output.c is what I'm using everywhere else.

However, as the number of formatted output messages increases, I really want the two to be separate files.  So it might make sense to move pcmk__out_prologue and pcmk__out_epilogue into a new location, if you can think of a good name.  This also means all the messages can go into pcmk_output.c where the register function is, and then I won't have to pollute the namespace by exposing them in a header.

clumens replied:
Okay I think I have this looking the way I want.

kgaillot commented on tools/crm_simulate.c:
```diff
@@ -1039,16 +1052,17 @@ main(int argc, char **argv)
-            out->info(out, "%sTransition Summary:\n", pcmk_any_flags_set(data_set->flags, pe_flag_show_scores|pe_flag_show_utilization)
-                      || options.modified ? "\n" : "");
+            PCMK__OUTPUT_SPACER_IF(out, printed == pcmk_rc_ok);
+            out->begin_list(out, NULL, NULL, "Transition Summary");
```
This change should be in the "is_quiet" commit

kgaillot commented on lib/pacemaker/pcmk_sched_native.c:
```diff
@@ -1473,7 +1473,7 @@ allowed_nodes_as_list(pe_resource_t *rsc, pe_working_set_t *data_set)
 
-    if (pcmk_is_set(data_set->flags, pe_flag_stdout)) {
+    if (data_set->priv != NULL || pcmk_is_set(data_set->flags, pe_flag_stdout)) {
         allowed_nodes = g_list_sort(allowed_nodes, sort_node_uname);
```
Since you set priv in the daemons, this will now always be done ... the goal was to avoid the overhead in the daemons

clumens replied:
I think the plan here was that I wanted to make sure things got logged, but that can be done without sorting.

clumens replied:
I've fixed this so the sorting no longer gets called, though I'm not thrilled about how I'm doing it.

kgaillot commented on tools/crm_simulate.c:
```diff
@@ -1030,11 +1031,25 @@ main(int argc, char **argv)
+            logger_out = pcmk__new_logger();
+            if (logger_out == NULL) {
+                crm_err("Can't log resource details due to internal error: %s\n",
+                        pcmk_rc_str(rc));
```
This message is done in pcmk__new_logger()

kgaillot commented on lib/pacemaker/pcmk_sched_allocate.c:
```diff
@@ -1199,12 +1199,14 @@ sort_rsc_process_order(gconstpointer a, gconstpointer b, gpointer data)
                                           pe_weights_forward | pe_weights_init);
-    pe__show_node_weights(true, NULL, resource2->id, r2_nodes);
+    pe__show_node_weights(true, NULL, resource2->id, r2_nodes,
+                          resource1->cluster->priv);
```
for consistency this should probably be resource2-> though it will be the same

kgaillot commented on include/crm/pengine/internal.h:
```diff
@@ -365,11 +365,12 @@ extern void pe__output_node(pe_node_t * node, gboolean details, pcmk__output_t *
                               int line, bool to_log, pe_resource_t *rsc,
-                              const char *comment, GHashTable *nodes);
+                              const char *comment, GHashTable *nodes,
+                              pcmk__output_t *out);
```
I'm thinking this should take the data set instead of out to simplify the callers

clumens replied:
It doesn't appear to do much to simplify the callers - I think it just replaced the pcmk__output_t parameter with a pe_working_set_t parameter.  But sure, easy enough to do.

kgaillot replied:
Yep that's what I meant, data_set instead of data_set->priv, to make it more object-oriented from the caller's point of view.

kgaillot reviewed (commented):
First round

clumens commented on tools/crm_simulate.c:
```diff
@@ -877,16 +886,16 @@ build_arg_context(pcmk__common_args_t *args, GOptionGroup **group) {
-    pcmk__add_arg_group(context, "output", "Output Options:",
-                        "Show output options", output_entries);
+    pcmk__add_arg_group(context, "artifact", "Artifact Options:",
+                        "Show artifact options", artifact_entries);
```
I'm not at all sold on the "Artifact" name.  It was just the best thing I could come up with in under a minute.

I kind of like having the output options group be just the formatted output options.  This is how all the other tools are at the moment, and I like that consistency.

kgaillot replied:
Makes sense. I can't think of anything better either.

clumens commented on tools/crm_simulate.c:
```diff
@@ -365,13 +358,16 @@ static GOptionEntry source_entries[] = {
-        quiet_log(" + Setting effective cluster time: %s", use_date);
+        if (!out->is_quiet(out)) {
+            out->info(out, "Setting effective cluster time: %s", use_date);
+        }
```
This would be nice, but there's a couple places on master that just have to be different.  For example:
```
    if (!out->is_quiet(out)) {
        out->info(out, "Status of %s: '%s' %s %s", crm_str(sys_from),
                  crm_str(state), (!pcmk__str_empty(last_updated))?
                  "last updated":"", crm_str(last_updated));
    } else {
        out->info(out, "%s", crm_str(state));
    }
```

If we can decide on a strategy for the three or four locations like that (`grep -r is_quiet -A 1 *`), I'm all for this.

kgaillot replied:
I'm thinking custom messages.

BTW ack will make your life a lot easier than grep :)

clumens replied:
There are a lot of places where out->info is called, most prominently in crm_resource and crm_mon.  Previously, I had been using out->info for stuff like various status messages, explaining that we're reconnecting to the daemons, etc.  Doing this will cause all those messages to go away, though if the user adds -Q that's probably what they want.  It might also be worth revisiting the purpose of all those uses of out->info.  Maybe some of them should turn into warnings, or just go away altogether.

kgaillot replied:
> Doing this will cause all those messages to go away, though if the user adds -Q that's probably what they want.

That's what gave me the idea. As a general rule, the existing implementation of --quiet across various tools dispenses with all output except some essential result. Having info() automatically respect it could simplify a lot of that and make the tools behave consistently. Not all tools currently support --quiet, but I think we're moving in that direction.

kgaillot replied:
It does mean we need custom messages for any usage that just produces less verbose output for quiet, but maybe that's worth it

clumens commented:
Updated various things.  Note the last patch - I'm not sure whether this is removing things that should be in the scores files that should be in there, or what.

clumens commented:
Updated again, fixed some problems with printing out promotion scores (by adding another output message).  The first several patches are going to get broken out into their own PR just as soon as I feel like I'm not going to have to add more.  Still a ways to go here.

kgaillot commented on lib/common/output_log.c:
```diff
@@ -82,12 +89,16 @@ log_reset(pcmk__output_t *out) {
 log_version(pcmk__output_t *out, bool extended) {
+    private_data_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
```
If we're going to assert this we may as well assert out != NULL && out->priv != NULL before grabbing priv

clumens replied:
Fixed, and I'll also go through the rest of the formatted output code to make sure this is being done consistently.  I'll add another patch to #2326 if necessary.

kgaillot commented on lib/common/output_log.c:
```diff
@@ -277,3 +291,11 @@ pcmk__mk_log_output(char **argv) {
+    private_data_t *priv = out->priv;
+
+    CRM_ASSERT(priv != NULL);
+    priv->log_level = log_level;
```
It occurs to me that with these formatter-specific utility functions, we should do nothing if the format isn't what it's supposed to be, just in case we call it somewhere with a generic output object by mistake (don't want to clobber some other format's private data)

clumens replied:
There's no better way to do this than by comparing strings (there's compiler extensions like typeof, but...), and some care does need to be taken since the XML functions can be used by HTML functions, but added.

kgaillot reviewed (commented):
I think the first 4 commits are basically ready, if you want to break them out

kgaillot commented:
> I think the first 4 commits are basically ready, if you want to break them out

... which you were obviously doing as I was reviewing them :)

kgaillot commented:
retest this please

kgaillot commented:
> retest this please

whoops, wrong PR :)

clumens commented:
This PR is getting tough to follow.  Rebased on top of master and the latest in #2326.

clumens commented:
I don't have anything else to add on this PR, so it's just going to be reacting to review comments.  It would definitely be worth looking at the last two patches to see if the changes to output are acceptable, because that's going to be the part that's going to be the most visible to users (and whatever might be scraping crm_simulate output).  I've verified that all existing tests pass, and all the output from crm_simulate validates against the new schema.

kgaillot commented on lib/pacemaker/pcmk_sched_allocate.c:
```diff
@@ -2868,12 +2870,30 @@ stage7(pe_working_set_t * data_set)
+    out = pcmk__new_logger();
+    if (out == NULL) {
+        return FALSE;
+    }
```
This previously logged at notice level, so I think we need to set that for this loop

clumens replied:
Fixed.

kgaillot commented on lib/pacemaker/pcmk_sched_messages.c:
```diff
@@ -36,13 +36,17 @@ log_resource_details(pe_working_set_t *data_set)
+        logger_out = pcmk__new_logger();
+        if (logger_out == NULL) {
+            return;
+        }
```
Nothing wrong with this approach, but I noticed crm_simulate directly calls LogNodeActions() when not quiet. If we're lucky, maybe we can just use the data_set output object here, and get rid of the direct call in crm_simulate.

clumens replied:
Making these changes has the following side effects:

* .scores files now have node and resource information written to them, like so:
```
--- a/cts/scheduler/scores/1-a-then-bm-move-b.scores
+++ b/cts/scheduler/scores/1-a-then-bm-move-b.scores
@@ -1,4 +1,6 @@
 
+A      (ocf:heartbeat:Dummy):   Started 18node1
+B      (ocf:heartbeat:Dummy):   Started 18node2
 pcmk__native_allocate: A allocation score on 18node1: 0
 pcmk__native_allocate: A allocation score on 18node2: 0
 pcmk__native_allocate: A allocation score on 18node3: 0
```

* .summary files have fencing messages removed and spacing has changed elsewhere, like so:
```
--- a/cts/scheduler/summary/year-2038.summary
+++ b/cts/scheduler/summary/year-2038.summary
@@ -46,11 +46,10 @@ Current cluster status:
       * openstack-cinder-volume-docker-0       (ocf:heartbeat:docker):  Started controller-0
 
 Transition Summary:
-  * Fence (reboot) overcloud-novacompute-1 'remote connection is unrecoverable'
-  * Stop       overcloud-novacompute-1              (            controller-1 )   due to node availability
+  * Stop       overcloud-novacompute-1              (            controller-1 )  due to node availability
   * Start      ip-10.0.0.110                        (            controller-1 )
   * Recover    stonith-fence_compute-fence-nova     (            controller-2 )
-  * Stop       compute-unfence-trigger:1            ( overcloud-novacompute-1 )   due to node availability
+  * Stop       compute-unfence-trigger:1            ( overcloud-novacompute-1 )  due to node availability
```

These are differences from my patch updating the output of cts/cts-scheduler, NOT from master.  So, I am wondering if those fencing messages ever should have been there.  The whitespace I'm not concerned about.  And then the node and resource stuff in the .scores files... that might point to a bug elsewhere.  I have not checked how this would look on a running system.  That's the next step.

clumens replied:
The fencing messages coem from that call to LogNodeActions in crm_simulate.c, around line 1073.  So I don't see how we could get rid of that.

clumens replied:
And then, the resources in the .scores file come from the call to log_resource_details after stage0 in pcmk__schedule_actions.  That can be taken care of by just adding a check for pe_flag_print not being set, since I think we only want to do that logging if we're running a daemon.  There's no point in logging if crm_simulate is being run.  So, that will at least let me get rid of the new logger setup in log_resource_details.

kgaillot replied:
There's a pcmk__is_daemon global boolean

clumens replied:
I refuse to be excited about the existence of a global variable, but I think it's going to be useful here.  I think it can probably eliminate the new pe_flag_print flag that I added.

kgaillot reviewed (commented):
This is not a full review but it's most of it

clumens commented on tools/crm_simulate.c:
```diff
@@ -77,6 +78,14 @@ extern gboolean bringing_nodes_online;
 #define INDENT "                                   "
 
+static pcmk__supported_format_t formats[] = {
+    PCMK__SUPPORTED_FORMAT_LOG,
```
Fixed.

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -10,6 +10,7 @@
 #include <crm_internal.h>
 #include <crm/common/results.h>
 #include <crm/common/output_internal.h>
+#include <crm/common/xml_internal.h>
```
Fixed.

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -846,21 +847,422 @@ node_action_xml(pcmk__output_t *out, va_list args)
+    { "pseudo-action", "default", pseudo_action },
+    { "pseudo-action", "xml", pseudo_action_xml },
+    { "rsc-action", "default", rsc_action },
+    { "rsc-action", "xml", rsc_action_xml },
```
Fixed.

clumens commented on lib/pacemaker/pcmk_sched_allocate.c:
```diff
@@ -3050,12 +3070,9 @@ LogNodeActions(pe_working_set_t * data_set, gboolean terminal)
+            pcmk__output_t *out = data_set->priv;
+            out->message(out, "node-action", task, node_name, action->reason);
         } else {
             crm_notice(" * %s %s\n", task, node_name);
```
Agreed, fixed.

kgaillot commented:
also bump feature set minor-minor version for new arguments

clumens commented:
> also bump feature set minor-minor version for new arguments

Added in a new patch.

clumens commented on tools/crm_simulate.c:
```diff
@@ -1028,6 +1047,7 @@ main(int argc, char **argv)
 
         cleanup_calculations(data_set);
         data_set->input = input;
+        data_set->priv = out;
```
Yes, if we are showing neither scores nor utilizations (which happens in cts/cts-cli) then data_set->priv will be set to the new logger_out here.  Afterwards, we'll need to set it back.  Otherwise, segfaults.

clumens commented on daemons/fenced/pacemaker-fenced.c:
```diff
@@ -1546,12 +1556,31 @@ main(int argc, char **argv)
+        crm_err("Can't log resource details due to internal error: %s\n",
+                pcmk_rc_str(rc));
+        crm_exit(CRM_EX_FATAL);
+    }
```
If this fails, out (and fenced_data_set->priv) will be NULL, which is going to trip up somewhere.  It'll either hit a CRM_ASSERT or a NULL pointer dereference we're not guarding because we assume it'll always be fine.  We could protect the code immediately following this, but I bet there'll be something else.

On the other hand, I can't see why pcmk__output_new would fail unless something was seriously wrong.  We control that the formatters will be registered ahead of time, and we pass a static string that is the formatter we want to use.

kgaillot replied:
Makes sense

clumens commented on tools/crm_simulate.c:
```diff
@@ -1010,28 +1003,31 @@ main(int argc, char **argv)
+                      "from further action due to failure",
                       data_set->disabled_resources, data_set->ninstances,
                       data_set->blocked_resources);
+            printed = pcmk_rc_ok;
```
Yeah, I like info returning a useful error code now.  #2332 opened for that, and I'll make the changes in this PR too.

kgaillot commented on tools/crm_simulate.c:
```diff
@@ -1010,28 +1003,31 @@ main(int argc, char **argv)
 
-        quiet_log("\nCurrent cluster status:\n");
+        PCMK__OUTPUT_SPACER_IF(out, printed == pcmk_rc_ok);
+        out->begin_list(out, NULL, NULL, "Current Cluster Status");
```
pcs currently looks for just two literal strings in crm_simulate output, "Current cluster status:" and "Transition Summary:". Let's leave them the same, maybe with a comment as to why it's different.

https://github.com/ClusterLabs/pcs/blob/7d04456984ef5c49edcffb799107732402877e1d/pcs/resource.py#L3195

clumens replied:
Good catch, fixed.

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -577,6 +578,274 @@ digests_xml(pcmk__output_t *out, va_list args)
+        reason = crm_strdup_printf(" due to %s", source->reason);
+
+    } else if (!pcmk_is_set(action->flags, pe_action_runnable)) {
+        reason = strdup(" blocked");
```
I've removed the spaces from the start of the reason strings.  The war on misplaced spaces continues!

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -596,6 +867,8 @@ static pcmk__message_entry_t fmt_functions[] = {
     { "stacks-constraints", "default", stacks_and_constraints },
     { "stacks-constraints", "xml", stacks_and_constraints_xml },
+    { "transition", "default", transition_default },
+    { "transition", "xml", transition_xml },
```
There's already a rsc_action, created by the inject-rsc-action.  I'm thinking about renaming all the XML elements created by inject-whatever messages to inject_whatever for clarity, even though they're encapsulated within an element which would allow telling them apart.  Doing so might then make renaming this one to rsc_action be okay.

clumens replied:
Fixed.

clumens commented on lib/pacemaker/pcmk_sched_native.c:
```diff
@@ -2580,38 +2475,37 @@ LogActions(pe_resource_t * rsc, pe_working_set_t * data_set, gboolean terminal)
+                out->message(out, "transition", "Re-promote", rsc, current, next, promote, demote);
             } else {
                 pe_rsc_info(rsc, "Leave   %s\t(%s %s)", rsc->id,
                             role2text(rsc->role), next->details->uname);
```
This output message is used in so many different ways that I think it's going to be tricky to merge this in.  At the least, I think I'll have to reproduce both the "eise if (start == NULL)..." condition as well as the "if (rsc->role == rsc->next_role)" condition surrounding it.  I'm fine doing that, if you don't think it's too much duplication.

kgaillot replied:
Not at all, that makes sense. The whole point of the message is to say what has to happen to get from the current role to the next role. Really, the entire LogActions() function starting with the first pe_rsc_info() call is the logical message.

clumens replied:
It might make sense to rearrange the code to reflect that, then.  The current transition_default and transition_xml functions (which need to be renamed, but whatever) could turn into helper functions that assemble the list_item out of all the various little pieces of data.  The bulk of the existing LogActions function then gets turned into the actual output message.

clumens replied:
And done.  I think I like the rearranged code better.  Next step is to rename all this transition stuff.

clumens commented:
I'm going to push what I've got, addressing most of the problems here, before I start doing a little rearranging when it comes to the transition stuff.  That way, I've got a backup.  I don't expect it'll change the code a ton.

clumens commented on tools/crm_simulate.c:
```diff
@@ -379,7 +379,7 @@ get_date(pe_working_set_t *data_set, bool print_original, char *use_date)
                             crm_time_log_date|crm_time_log_timeofday);
 
-            printf("Using the original execution date of: %s\n", when);
+            out->info(out, "Using the original execution date of: %s", when);
```
Yeah maybe, but it's tricky.  get_date is called in a lot of different places - for each profile run, before the biggest piece of work in main, after performing modifications, and when printing out the revised cluster status.  It's not clear to me how most of those fit in to the XML schema.

kgaillot replied:
Yeah, maybe leave for another day

kgaillot commented on include/crm/pengine/pe_types.h:
```diff
@@ -116,6 +116,9 @@ enum pe_find {
 #  define pe_flag_sanitized             0x00200000ULL
+
+#  define pe_flag_print                 0x00400000ULL
+//! \deprecated
```
pcmk__is_daemon might be suitable for this, but all the other tools will hit !pcmk__is_daemon as well so make sure they don't suddenly become noisy :)

clumens replied:
I haven't seen any extra noisiness yet, but I'll stay on the lookout.

clumens replied:
I haven't seen any extra noisiness yet, but I'll stay on the lookout.

kgaillot commented on lib/pacemaker/pcmk_sched_allocate.c:
```diff
@@ -280,11 +280,19 @@ check_action_definition(pe_resource_t * rsc, pe_node_t * active_node, xmlNode *
+                       " on %s changed: %s\n",
+                       rsc->id, task, interval_ms, active_node->details->uname,
+                       crm_element_value(xml_op, XML_ATTR_TRANSITION_MAGIC));
+            }
```
This is all internal API so I don't think we have to care about when data_set->priv is NULL

kgaillot commented on lib/pengine/utils.c:
```diff
@@ -2079,8 +2079,14 @@ pe_fence_op(pe_node_t * node, const char *op, bool optional, const char *reason,
+                                      node->details->uname, match->id);
+                        } else {
+                            fprintf(stdout, "  notice: Unfencing %s (remote): because the definition of %s changed\n", node->details->uname, match->id);
+                        }
```
This is public API but I don't think it matters whether we print anything when priv is NULL, the output is not central to the purpose (as public API)

kgaillot reviewed (commented):
A few minor comments on the current state

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -102,6 +106,232 @@ do_locations_list_xml(pcmk__output_t *out, pe_resource_t *rsc, bool add_header)
+        rsc_width = len + 2;
+    }
+
+    if(rsc->role > RSC_ROLE_STARTED || rsc->next_role > RSC_ROLE_SLAVE) {
```
This is all just cruft from moving the code.  Fixed.

clumens replied:
Whoops - rsc->role vs. rsc->next_role here.  Tricky stuff.

kgaillot replied:
Doh, sorry

clumens commented on lib/pacemaker/pcmk_output.c:
```diff
@@ -579,6 +809,241 @@ digests_xml(pcmk__output_t *out, va_list args)
+        mk_list_item = do_rsc_action_item_xml;
+    } else {
+        mk_list_item = do_rsc_action_item;
+    }
```
It could be.  I was trying to keep the amount of code duplication to a minimum.  There's a lot of stuff in rsc_action_default that needs to be done no matter what output format, so I wanted to share all that code.  I guess that could still be done and turn do_rsc_action_item and do_rsc_action_item_xml into the custom messages.

kgaillot commented on lib/common/crmcommon_private.h:
```diff
@@ -276,7 +276,6 @@ pcmk__ipc_methods_t *pcmk__pacemakerd_api_methods(void);
  * It is set when logging is initialized, and does not need to be set directly.
  */
-G_GNUC_INTERNAL
 extern bool pcmk__is_daemon;
```
I didn't realize this was library-local, let's move this to include/crm/common/internal.h (or add an accessor function)

clumens replied:
Fixed.

kgaillot commented on lib/pacemaker/pcmk_sched_promotable.c:
```diff
@@ -753,8 +755,13 @@ pcmk__set_instance_roles(pe_resource_t *rsc, pe_working_set_t *data_set)
+                out->info(out, "%s promotion score on %s: %s",
+                          child_rsc->id,
+                          (chosen? chosen->details->uname : "none"), score);
+            } else {
```
I don't think we need to keep the else block here or in similar usages

clumens replied:
Fixed.

kgaillot reviewed (commented):
I think this is about ready to merge, let me know when you're happy with it

clumens commented:
> I think this is about ready to merge, let me know when you're happy with it

I don't have anything else that needs to be done.  I'm sure we're gonna find some bugs, but it's going to take getting this in wider usage to start finding them.
